internal/provider/provider.go

```diff
@@ -0,0 +1,253 @@
+	p.logger.Debug("vault: performing vault login...")
+
+	req := client.NewRequest("POST", "/v1/auth/"+vaultKubernetesMountPath+"/login")
+	err := req.SetJSONBody(struct {
+		Role string `json:"role"`
+		JWT  string `json:"jwt"`
+	}{
+		roleName,
+		jwt,
+	})
+	if err != nil {
+		return "", err
+	}
+	secret, err := vaultclient.Do(ctx, client, req)
+	if err != nil {
+		return "", fmt.Errorf("failed to login: %w", err)
+	}
+
+	client.SetToken(secret.Auth.ClientToken)
+
+	return secret.Auth.ClientToken, nil
+}
+
+func (p *provider) getSecret(ctx context.Context, client *api.Client, secret config.Secret) (content string, err error) {
+	p.logger.Debug("vault: getting secrets from vault...")
+
+	s := regexp.MustCompile("/+").Split(secret.ObjectPath, 3)
+	if len(s) < 3 {
+		return "", fmt.Errorf("unable to parse secret path %q", secret.ObjectPath)
+	}
+	secretPrefix := s[1]
+	secretSuffix := s[2]
+
+	secretMountType, secretMountVersion, err := p.getMountInfo(ctx, client, secretPrefix)
+	if err != nil {
+		return "", err
+	}
+
+	endpoint, err := generateSecretEndpoint(secretMountType, secretMountVersion, secretPrefix, secretSuffix)
+	if err != nil {
+		return "", err
+	}
+
+	p.logger.Debug("vault: Requesting valid secret mounted", "endpoint", endpoint)
+
+	req := client.NewRequest("GET", endpoint)
+	if secret.ObjectVersion != "" {
+		req.Params = url.Values{
+			"version": []string{secret.ObjectVersion},
+		}
+	}
+	resp, err := vaultclient.Do(ctx, client, req)
+	if err != nil {
+		return "", fmt.Errorf("couldn't read secret %q: %w", secret.ObjectName, err)
+	}
+	if resp == nil || resp.Data == nil {
+		return "", fmt.Errorf("empty response from %q, warnings: %v", req.URL.Path, resp.Warnings)
+	}
+
+	var data map[string]interface{}
+	d, ok := resp.Data["data"]
+	if ok {
+		data, ok = d.(map[string]interface{})
+	}
+	if !ok {
+		data = resp.Data
+	}
+
+	content, ok = data[secret.ObjectName].(string)
+	if !ok {
+		return "", fmt.Errorf("failed to get secret content %q as string", data[secret.ObjectName])
+	}
+
+	return content, nil
+}
+
+// MountSecretsStoreObjectContent mounts content of the vault object to target path
+func (p *provider) MountSecretsStoreObjectContent(ctx context.Context, cfg config.Config) (map[string]string, error) {
+	versions := make(map[string]string)
+	for _, secret := range cfg.Parameters.Secrets {
+		content, err := p.getSecretContent(ctx, cfg.Parameters, secret)
+		if err != nil {
+			return nil, err
+		}
+		versions[fmt.Sprintf("%s:%s:%s", secret.ObjectName, secret.ObjectPath, secret.ObjectVersion)] = secret.ObjectVersion
+		err = writeSecret(p.logger, cfg.TargetPath, secret.ObjectName, content, cfg.FilePermission)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return versions, nil
+}
+
+func writeSecret(logger hclog.Logger, directory string, file string, content string, permission os.FileMode) error {
```

Comment:
Perhaps we should turn this into a method and have a secret struct? In the future there could be more things we need to add such as leases, templates, etc. Thoughts?

Reply:
I've given this some thought and played around a little with the code. I think you're right that something along these lines will be useful in the future, but I don't have a firm enough idea of how it will look to do it yet, so I'll defer for now.